New complete version of "Svelte SEO"

kudadam opened the pull request:
@artiebits ,
Finally, I am done with the proposed new version I suggested in #44.

It comes with new features plus some breaking changes, not many, I think one or two properties' names were renamed.
Here are the new features.
* Supports all open graph properties
* Includes additional properties
* Supports all twitter tags
* Fully typed

The coding style in this new version was changed a bit.
For example, in the old version, we used to use if statements to check if a property existed, 
before adding its meta tag but in this version, we loop through all the properties provided by the user.
You might ask what if the user provides a property which isn't specified?
Well, since the component is well-typed. The user will be notified of the error.

Also, for the breaking changes, I had to change some of the variable names. 
For example, `openGraph.article.publishedTime`  was changed to `openGraph.article.published_time`. 
The reason is simple, we now use a custom parser to convert camel case to the required case so every capital letter is placed by ":" + the letter since we are now using loops.

If we use `openGraph.article.publishedTime`, the parser will convert it to `article:published:time`, since this is not valid, I need to change some of the variable names.

Also, I started the project from a fresh sveltekit repository so I advice we work on this and fix any changes before we merge it as a new version.

Talking about version, since this change is going to include breaking changes, I suggest we name this version 2.0.0 before any other fixes are added.

Also, this project uses playwright tests to tests every single property.

If you have any more questions, feel free to ask me

artiebits commented:
hi @kudadam, thank you very much for your contribution, I appreciate it very much!

I will review your changes this weekend.

artiebits commented on .prettierrc:
```diff
@@ -0,0 +1,9 @@
+	"plugins": ["prettier-plugin-svelte"],
+	"pluginSearchDirs": ["."],
+	"overrides": [{ "files": "*.svelte", "options": { "parser": "svelte" } }]
+}
```
I would highly prefer to keep default Prettier config.

artiebits replied:
default Prettier config means removing .prettierrc file. Please remove it completely and run prettier again. thank you

artiebits commented on .eslintcache:
```diff
@@ -0,0 +1 @@
+[{"C:\\Projects\\svelte-seo\\src\\lib\\index.svelte":"1","C:\\Projects\\svelte-seo\\src\\lib\\openGraph.svelte":"2","C:\\Projects\\svelte-seo\\src\\routes\\+page.svelte":"3"},{"size":3329,"mtime":1668633882001,"results":"4","hashOfConfig":"5"},{"size":3056,"mtime":1668633882671,"results":"6","hashOfConfig":"5"},{"size":501,"mtime":1668633882821,"results":"7","hashOfConfig":"5"},{"filePath":"8","m ...
```
please add this file to .gitignore

kudadam replied:
Done

artiebits replied:
Could you please delete this file from the repo, since it is in gitignore?

artiebits replied:
Could you please delete this file from the repo, since it is in gitignore?

kudadam replied:
@artiebits , it has been deleted

artiebits commented on .travis.yml:
```diff
@@ -1,18 +0,0 @@
-language: node_js
```
why CICD is deleted?

kudadam replied:
I have added it back

kudadam commented:
@artiebits  do look and tell me if there are any more changes to make

kudadam commented:
@artiebits , am still waiting on your review about the pull request

artiebits commented on .husky/.gitignore:
```diff
@@ -0,0 +1 @@
+_
```
what is the purpose of this file?

kudadam replied:
@artiebits , the .prettierrc file wasn't generated by me. It was generated by Sveltekit, do you still want me to delete it??

kudadam replied:
Also, the .husky/.gitignore was generated by husky itself, I used it's latest versions

artiebits replied:
I would prefer to keep the previous prettier config (default one). It will also make it easier to review your changes because changing code formatting changed every single file

artiebits replied:
hi @kudadam would you be okay to keep the previous prettier formatting? It would help to finalise this PR. Thanks!

kudadam replied:
@artiebits , I can't find the default prettier formatting in your repo. Can you show me where it is so I will fix it in?

artiebits replied:
just remove the existing prettier config file, then Prettier will apply its default config

kudadam replied:
@artiebits , It has been deleted

artiebits replied:
Hey @kudadam. Happy New Year and best wishes for the year ahead!

Since the custom prettier config deleted, will you be able to run Prettier on all files, so the default config applies? 

Sorry for taking so long to reply.

kudadam replied:
@artiebits , Happy New Year to you too, I will start working on it

artiebits reviewed (changes requested):
sorry for the delay in response. Could you please make these changes?

kudadam commented:
@artiebits , do you want me to remove the default prettier configuration file??

kudadam commented:
@artiebits, the default prettier settings has been ran on the whole code base. Am still working on the readme though

artiebits commented:
When I do `nom run check` it throws errors:
```
Error: Cannot find name 'SvelteSeo'.

  /**@type {SvelteSeo['twitter']} */
  twitter: {


/Users/arturkhusaenov/code/personal/svelte-seo/src/routes/index.js:53:13
Error: Cannot find name 'SvelteSeo'.

  /**@type {SvelteSeo['openGraph']} */
  openGraph: {


====================================
svelte-check found 2 errors and 0 warnings
```

when I run `npm run test` it throws:
```
[WebServer] /bin/sh: pnpm: command not found
Error: Process from config.webServer was not able to start. Exit code: 127

82 skipped
1 error was not a part of any test, see above for details
```

kudadam commented:
Do you have pnpm installed globally?

artiebits commented:
Do you think it would be possible to create pages for article, video, profile, etc, like here https://github.com/garmeeh/next-seo/tree/master/e2e/pages? It might be better then adding all tags into one single file. 

The rest looks fine to me. if we fix these final things than I would love to merge your PR.

And sorry for keeping you waiting like this!

kudadam commented:
Alright, I will do that tonight

artiebits commented:
> Do you have pnpm installed globally?

will current travisCI work with pnpm or it needs to be updated?

kudadam commented:
It needs to be updated

kudadam commented:
Is travis used to run the tests??

artiebits commented on package.json:
```diff
@@ -1,61 +1,48 @@
+    "*.svelte": "eslint --cache --fix",
+    "*.{css,md,svelte}": "prettier --write",
+    "*.js": "eslint --cache --fix",
+    "*.{js,css,md}": "prettier --write"
```
why like this?

why not?
```
    "*.{js,svelte}": "eslint --cache --fix",
    "*.{js,md,svelte}": "prettier --write"
```

kudadam replied:
Hope you've heard of `npm create svelte@latest`? When you create a default SvelteKit  project, it includes those by default.

artiebits replied:
and why you didn't add ts file?

kudadam replied:
> why some files are js, and some are ts?

There are no ts files in the repo

kudadam replied:
Why would I add a ts file?

artiebits replied:
> Hope you've heard of `npm create svelte@latest`? When you create a default SvelteKit project, it includes those by default.

okay, but there is no css in this repo and I was really curious why you have to do 

"*.svelte": "eslint --cache --fix",
"*.js": "eslint --cache --fix"

insted of 

 "*.{js,svelte}": "eslint --cache --fix"

let's leave it

artiebits replied:
> There are no ts files in the repo

strange, I see 4 ts files in your pull request

kudadam replied:
First of all, I have changed the eslint to reflect your changes.
The only other typescript files are the one generated by SvelteKit

artiebits replied:
even if a file generated by SvelteKit it should be linted. but you can leave it, I will add missing files (ts, json) later myself

artiebits commented:
why some files are js, and some are ts?

kudadam commented:
> Do you think it would be possible to create pages for article, video, profile, etc, like here https://github.com/garmeeh/next-seo/tree/master/e2e/pages? It might be better then adding all tags into one single file.
> 
> The rest looks fine to me. if we fix these final things than I would love to merge your PR.
> 
> And sorry for keeping you waiting like this!

Upon careful consideration, it wouldn't be a good idea.
I know you were inspired by that repo when creating this but trying to emulate their everything wouldn't help.
The less the code, the fewer the bugs, this method already works so why should we split them??
Also, if you are to split them all into equal files, it would mean more test files to write to accomplish what is already accomplished now. 
The reason there is only one file is to unify everything for a single source of truth. I do not know if this is convincing you but it's the best. Let's try to reduce the code we write so that maintaining the code will be much easier.

If you create a page for each open graph tag, then you would also have to create pages for dynamic twitter tags, then you will also create pages for different JSON-LD's, it's just going to bloat the codebase.

Another issue is that, you are barely around for this repo. I opened this issue months ago yet it does not seem to get anywhere closer to completion. The main reason why I opened this pull request was to add improvements to your module which it was lacking. I had a project which needed these changes but because of your delay. I had to look for an alternative. If you wouldn't always be around for your code, you could at least give maintainers access to the repo so important bugs and fixes can be made.

![image](https://user-images.githubusercontent.com/63173187/213218628-2ed5323d-a678-4b95-97e4-73cc4e739b6d.png)

These are the trends between this plugin and another alternative. It's getting more downloads. Development on this repo is also uncertain, I don't know if you are getting me but trying to make everything smooth for anyone who wants to contribute to your code.

artiebits commented:
> > Do you think it would be possible to create pages for article, video, profile, etc, like here https://github.com/garmeeh/next-seo/tree/master/e2e/pages? It might be better then adding all tags into one single file.
> > The rest looks fine to me. if we fix these final things than I would love to merge your PR.
> > And sorry for keeping you waiting like this!
> 
> Upon careful consideration, it wouldn't be a good idea. I know you were inspired by that repo when creating this but trying to emulate their everything wouldn't help. The less the code, the fewer the bugs, this method already works so why should we split them?? Also, if you are to split them all into equal files, it would mean more test files to write to accomplish what is already accomplished now. The reason there is only one file is to unify everything for a single source of truth. I do not know if this is convincing you but it's the best. Let's try to reduce the code we write so that maintaining the code will be much easier.
> 
> If you create a page for each open graph tag, then you would also have to create pages for dynamic twitter tags, then you will also create pages for different JSON-LD's, it's just going to bloat the codebase.
> 
> Another issue is that, you are barely around for this repo. I opened this issue months ago yet it does not seem to get anywhere closer to completion. The main reason why I opened this pull request was to add improvements to your module which it was lacking. I had a project which needed these changes but because of your delay. I had to look for an alternative. If you wouldn't always be around for your code, you could at least give maintainers access to the repo so important bugs and fixes can be made.
> 
> ![image](https://user-images.githubusercontent.com/63173187/213218628-2ed5323d-a678-4b95-97e4-73cc4e739b6d.png)
> 
> These are the trends between this plugin and another alternative. It's getting more downloads. Development on this repo is also uncertain, I don't know if you are getting me but trying to make everything smooth for anyone who wants to contribute to your code.

hi Lucretius,

Good point, let's not make this PR bigger, since it's already big.

Thanks for your contribution, and I understand your frustration about time it takes to review your PR, but please understand me too. Before you started working on it I asked to keep changes as small as possible so it would be easier for me to review them. However, instead of rewriting this package step by step you open one big PR, and changed the formatting which touched every single file.

Let's try to fix merge conflicts, CICD, errors in console, and I think we are good to go! Again, thank you very much for your contribution, not that much left, let's finish it and merge to mater

kudadam commented:
The testing, why don't you use github actions??
Wouldn't that be simpler??

artiebits commented:
Hi @kudadam these are the only things that left. let me know if I can help you with some of these things

1. `yarn run check` and `npm run check` throws errors:
```
Error: Cannot find name 'SvelteSeo'.

  /**@type {SvelteSeo['twitter']} */
  twitter: {

/Users/arturkhusaenov/code/personal/svelte-seo/src/routes/index.js:53:13
Error: Cannot find name 'SvelteSeo'.

  /**@type {SvelteSeo['openGraph']} */
  openGraph: {

====================================
svelte-check found 2 errors and 0 warnings
```

2. `yarn test` and `npm run test` throws:
```
[WebServer] /bin/sh: pnpm: command not found
Error: Process from config.webServer was not able to start. Exit code: 127

82 skipped
1 error was not a part of any test, see above for details
```

3. update travis.yml to use new commands
add these 
```
- npm run build
- npm run check
- npm run test 
- npm run lint
```
remove
```
 # install dependencies and check that Cypress can run
  override:
    - npm ci
    - npm run cy:verify
```

4. resolve the conflict in yarn.lock. I guess you can remove it and do `yarn` again

artiebits commented:
> Do you have pnpm installed globally?

I do, pnpm throws those errors too

kudadam commented:
> Hi @kudadam these are the only things that left. let me know if I can help you with some of these things
> 
> 1. `yarn run check` and `npm run check` throws errors:
> 
> ```
> Error: Cannot find name 'SvelteSeo'.
> 
>   /**@type {SvelteSeo['twitter']} */
>   twitter: {
> 
> /Users/arturkhusaenov/code/personal/svelte-seo/src/routes/index.js:53:13
> Error: Cannot find name 'SvelteSeo'.
> 
>   /**@type {SvelteSeo['openGraph']} */
>   openGraph: {
> 
> ====================================
> svelte-check found 2 errors and 0 warnings
> ```
> 
> 2. `yarn test` and `npm run test` throws:
> 
> ```
> [WebServer] /bin/sh: pnpm: command not found
> Error: Process from config.webServer was not able to start. Exit code: 127
> 
> 82 skipped
> 1 error was not a part of any test, see above for details
> ```
> 
> 3. update travis.yml to use new commands
>    add these
> 
> ```
> - build
> - check
> - test 
> - lint
> ```
> 
> remove
> 
> ```
>  # install dependencies and check that Cypress can run
>   override:
>     - npm ci
>     - npm run cy:verify
> ```
> 
> 4. resolve the conflict in yarn.lock. I guess you can remove it and do `yarn` again

The reason why number 2 is not working is because, it's set to use pnpm. It should be `pnpm test`.

Hey, quick question, which package manager should we use, pnpm, npm or yarn.

kudadam commented:
Talking of travis, why don't you use Github actions for the tests??
I can see you use it for publishing the package so it would be more nice to just unify the whole thing or what do you think

artiebits commented:
> The reason why number 2 is not working is because, it's set to use pnpm. It should be `pnpm test`.
> 
> Hey, quick question, which package manager should we use, pnpm, npm or yarn.
>
> Talking of travis, why don't you use Github actions for the tests??

1. both `pnpm run test` and `pnpm run` throw the same errors.
2. this package uses Yarn (there is yarn.lock), so I would prefer to stick with Yarn and don't change the package manager within this PR.
3. I've added Travis almost 3 years ago when GitHub actions was quite new. Travis works perfectly fine to run just a few command. I don't mind moving to GitHub actions but let's first finalise this PR, and then think of other improvements.

artiebits commented:
great, thanks. please resolve the conflict in yarn.lock so I will be able to merge your PR. I guess you can remove yarn.lock and do `yarn` again

kudadam commented:
I just tried, it didn't work

kudadam commented:
After all the tests passes, don't merge yet. Go through the readme to see if you would make any changes

kudadam commented:
I don't know how to debug Travis CI so you can at least tell me why the tests are failing

artiebits commented:
please try to downgrade to 16.14

kudadam commented:
The command "npm config set progress false" failed and exited with 1 during .
What does that mean

kudadam commented:
I think I know what is happening, the VM travis is trying to use can't install node 17. Yarn also  requires 18

artiebits commented:
> npm run lint didn't work. You need to run format before

it didn't work because not all files follow formatting rules. you need to run `npm run format` locally and push the changes.

having `npm run format` in CI doesn't make sense because then you can push your code without following formatting rules but it will always pass in CI.

kudadam commented:
Hurray🥂🎆🎆🎆🎉🎉🎉🎉

artiebits commented on .travis.yml:
```diff
@@ -1,18 +1,18 @@
-    - npm run cy:verify
 
 script:
+  - npm run format
```
you don't need `npm run format` in Travis config file, because there is no point in formatting files in CI. Because then you can push your code without following formatting rules but it will always pass in CI.
1. run `npm run format` locally and push the changes.
2. remove `npm run format` from this file

kudadam replied:
Has been fixed

artiebits commented on package.json:
```diff
@@ -1,61 +1,47 @@
-    "cy:run": "cypress run",
-    "cy:open": "cypress open",
-    "test": "run-p --race dev cy:run",
-    "prepublishOnly": "yarn build"
```
please return this line back, needed it to publish a new version of the package

kudadam replied:
but it doesn't use cypress to run the tests. Forgotten??

artiebits replied:
I am talking about `"prepublishOnly": "yarn build"` :)

artiebits commented:
I've published a beta version with your changes `svelte-seo@2.0.0-beta.0` let's test it and merge your contribution to master!

kudadam commented:
> I've published a beta version with your changes `svelte-seo@2.0.0-beta.0` let's test it and merge your contribution to master!

alright

kudadam commented:
However, there are some breaking changes I think we need to specify

artiebits commented:
hi @kudadam I am testing `svelte-seo@2.0.0-beta.0` and getting `Failed to resolve entry for package "svelte-seo". The package may have incorrect main/module/exports specified in its package.json.` Does it work for you?

artiebits commented on package.json:
```diff
@@ -1,61 +1,47 @@
-  "main": "dist/index.min.js",
-  "types": "types/index.d.ts",
-  "repository": "git@github.com:artiebits/svelte-seo.git",
+  "version": "0.0.1",
```
should be "version": "1.4.1"

artiebits commented on README.md:
```diff
@@ -2,232 +2,203 @@
-<!-- START doctoc generated TOC please keep comment here to allow auto update -->
-<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
-
-- [Installing](#installing)
```
please generate the TOC using doctoc

artiebits commented on README.md:
```diff
@@ -2,232 +2,203 @@
-
-## License
-
-MIT
```
please return Acknowledgements and License sections

artiebits commented on README.md:
```diff
@@ -2,232 +2,203 @@
-    image: "https://www.example.com/images/cover.jpg",
-    imageAlt: "Alt text for the card!",
-  }}
+<Head
```
please use SvelteSeo name instead of head

artiebits commented on .travis.yml:
```diff
@@ -1,18 +1,17 @@
 script:
+  - npm run lint
+  - npm run check
+  - npx playwright install
```
are you sure `npx playwright install` should be in scripts section, or `before_install` or something like this? please investigate it

artiebits commented on src/tests/seo.test.js:
```diff
@@ -0,0 +1,89 @@
+  });
+
+  test("Loads the title", async ({ page }) => {
+    await expect(page).toHaveTitle(data?.title);
```
please use the exact text expected in the result. for example, `await expect(page).toHaveTitle("Hello World");`

kudadam commented:
> hi @kudadam I am testing `svelte-seo@2.0.0-beta.0` and getting `Failed to resolve entry for package "svelte-seo". The package may have incorrect main/module/exports specified in its package.json.` Does it work for you?

Yes, I'm getting the same error

artiebits commented:
> When you run `yarn run build`. SvelteKit packages the package into a folder name `package` inside the root directory. That is what you need to publish

you are right. there was a configuration in package.json but you deleted it:
```
  "module": "dist/index.min.mjs",
  "main": "dist/index.min.js",
  "types": "types/index.d.ts",
```
please add the relevant configuration to the package.json and test it by building and importing the package to some test project locally

kudadam commented:
> > When you run `yarn run build`. SvelteKit packages the package into a folder name `package` inside the root directory. That is what you need to publish
> 
> you are right. there was a configuration in package.json but you deleted it:
> 
> ```
>   "module": "dist/index.min.mjs",
>   "main": "dist/index.min.js",
>   "types": "types/index.d.ts",
> ```
> 
> Please add the relevant configuration to the package.json and test it by building and importing the package to some test project locally.
> 
> I will be on vacation until February 8th! see you later

No @artiebits, you got it all wrong, I am not supposed to add these scripts to the package.json. When you run `yarn build`. It runs a command called `svelte-package` which is different from the command which builds svelte apps.
It generates a folder called package. Inside the generated folder, it contains another `package.json` with the scripts above.
The problem was how you deployed the package.

![image](https://user-images.githubusercontent.com/63173187/215263334-77f1fe7c-fd06-46a9-838c-9bafd9a54118.png)
Even the `svelte-package` command says we should change into the package directory and publish it instead of the root repository.

![image](https://user-images.githubusercontent.com/63173187/215263362-26f55711-a068-4f3b-81d4-4007f8d5a18f.png)

This is how the package directory looks like. 

![image](https://user-images.githubusercontent.com/63173187/215263380-3a4347b2-983b-4eda-aeea-49c2df5d642c.png)

This is also how the package.json inside the package directory looks like. I am not supposed to add those scripts for it to work, besides, SvelteKit doesn't generate a folder like `dist` or something.
I have tried publishing the package folder and it worked. Try to see if you can do same and the issue would be resolved.
Happy vacation!

kudadam commented:
Hey, @artiebits, I'm glad everything went well and we're ready to publish. Before doing so, we also need to specify the breaking changes. 
I will list all the changes I made here.

**Changes** 
* Added `base` (Allows you to set the base URL)
* Added `applicationName` (Allows you to set the web app's name in the head)
* Added `themeColor` (Allows you to set the theme colour of the website)
* Added `nositelinkssearchbox` (Prevents google from showing a search box for your website.
* Added `notranslate` (Prevents google from translating your website)
* Added `amp` (Allows you to set the URL to the amp version of the page.
* Added `manifest` (Set URL to your website's manifest.json)
* Added `languageAlternates` (Set alternative language versions of the page
*

artiebits commented:
thank you for your contribution! https://github.com/artiebits/svelte-seo/releases/tag/v1.5.0
